Sources/LeafKit/LeafRenderer.swift

```diff
@@ -1,59 +1,145 @@
+    case unknownError(String)
+    case unsupportedFeature(String)
+    case cachingDisabled
+    case keyExists(String)
+    case noValueForKey(String)
+    case unresolvedAST(String, String)
+    case noTemplateExists(String)
+    case cyclicalReference(String, String)
+    case lexerError(LexerError)
+
+    var localizedDescription: String {
+        switch self {
+            case .unknownError(let message): return message
+            case .unsupportedFeature(let feature): return "\(feature) is not implemented"
+            case .cachingDisabled: return "Caching is globablly disabled"
+            case .keyExists(let key): return "Existing entry \(key): use insert with replace=true to overrride"
+            case .noValueForKey(let key): return "No cache entry exists for \(key)"
+            case .unresolvedAST(let key, let dependencies):
+                return "Flat AST expected; \(key) has unresolved dependencies: \(dependencies)"
+            case .noTemplateExists(let key): return "No template found named \(key)"
+            case .cyclicalReference(let key, let chain): return "\(key) was referenced in [\(chain)], causing a cyclical loop"
+            case .lexerError(let e): return e.localizedDescription
+        }
+    }
+
+    init(_ message: String) {
+        self = .unknownError(message)
+    }
+}
+
 public var defaultTags: [String: LeafTag] = [
     "lowercased": Lowercased(),
 ]
 
 public struct LeafConfiguration {
     public var rootDirectory: String
-    
+
     public init(rootDirectory: String) {
         self.rootDirectory = rootDirectory
     }
 }
 
 public protocol LeafCache {
+    // Superseded by insert with remove: parameter - Remove in Leaf-Kit 2?
+    @available(*, deprecated, message: "Use insert with replace parameter instead")
     func insert(
         _ document: ResolvedDocument,
         on loop: EventLoop
     ) -> EventLoopFuture<ResolvedDocument>
+    
+    func insert(
+        _ document: ResolvedDocument,
+        on loop: EventLoop,
+        replace: Bool
+    ) -> EventLoopFuture<ResolvedDocument>
+    
     func load(
         documentName: String,
         on loop: EventLoop
     ) -> EventLoopFuture<ResolvedDocument?>
+    
+    /// - return nil if cache entry didn't exist in the first place, true if purged
+    /// - will never return false in this design but should be capable of it
+    ///   in the event a cache implements dependency tracking between templates
+    func remove(
+        _ documentName: String,
+        on loop: EventLoop
+    ) -> EventLoopFuture<Bool?>
+    
+    func entryCount() -> Int
+    
     var isEnabled : Bool { get set }
 }
 
 public final class DefaultLeafCache: LeafCache {
     let lock: Lock
     var cache: [String: ResolvedDocument]
     public var isEnabled: Bool = true
-    
+
     public init() {
         self.lock = .init()
         self.cache = [:]
     }
     
+    // Superseded by insert with remove: parameter - Remove in Leaf-Kit 2?
```

Comment:
👋  Really nice work on this PR. Just doing a bit of drive by PR review.

I was wondering if there there is a reason to have this obsoleted method as callers can transparently call the `public func insert( _ document: ResolvedDocument, on loop: EventLoop, replace: Bool = false)` version if you remove this?

Reply:
Thanks! I would agree that functionally it's not necessary if someone's using LeafKit as-is but from a previous PR I was working on, I did my best to avoid any breaking changes should someone be using a different implementation instead of DefaultLeafCache

Reply:
Cool, so you should be able to safely delete the `insert(_:on)` method if you leave the `insert(_:on:replace)` method with the default parameter for replace.

`insert(_:on:replace)` gives you both versions, so you won't break clients by just having the longer version with default parameters. 👍

Reply:
I've run out of coffee and need to think about this 😂 100% agree that I'd like to remove it, think I was just leaving it as-is for the moment since behavior of inserting to the cache is ambiguous as to what it SHOULD do - currently DefaultLeafCache would just overwrite an existing entry, and someone *may* have taken advantage of that to force template updates?